File: Leetcode/DP/139.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
    #-------------1---------------
    #     word_set = set(wordDict)
    #     q = deque()
    #     visited = set()

    #     q.append(0)
    #     while q:
    #         start = q.popleft()
    #         if start in visited:
    #             continue
    #         for end in range(start + 1, len(s) + 1):
    #             if s[start:end] in word_set:
    #                 q.append(end)
    #                 if end == len(s):
    #                     return True
    #         visited.add(start)
    #     return False

    #-------------2---------------
        word_set = set(wordDict)
        dp = [False] * (len(s) + 1)
        dp[0] = True
        
        for i in range(1, len(s) + 1):
            for j in range(i):
                if dp[j] and s[j:i] in word_set:
                    dp[i] = True
                    break
        return dp[len(s)]

File: Leetcode/DP/139.js
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
var wordBreak = function (s, wordDict) {
  const dp = Array(s.length + 1).fill(false);
  dp[0] = true;

  for (let i = 1; i <= s.length; i++) {
    for (const word of wordDict) {
      const start = i - word.length;
      if (
        start >= 0 &&
        dp[start] &&
        s.substring(start, start + word.length) === word
      ) {
        dp[i] = true;
        break;
      }
    }
  }

  return dp[s.length];
};
